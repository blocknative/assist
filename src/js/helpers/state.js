export const initialState = {
  version: null,
  validApiKey: 'unknown',
  supportedNetwork: 'unknown',
  config: undefined,
  userAgent: null,
  mobileDevice: null,
  validBrowser: null,
  legacyWeb3: null,
  modernWeb3: null,
  web3Version: null,
  web3Instance: null,
  currentProvider: null,
  web3Wallet: null,
  legacyWallet: null,
  modernWallet: null,
  accessToAccounts: null,
  accountAddress: null,
  walletLoggedIn: null,
  walletEnabled: null,
  walletEnableCalled: null,
  walletEnableCanceled: null,
  accountBalance: null,
  minimumBalance: null,
  correctNetwork: null,
  userCurrentNetworkId: null,
  socket: null,
  pendingSocketConnection: null,
  socketConnection: null,
  transactionQueue: [],
  transactionAwaitingApproval: false,
  iframe: null,
  iframeDocument: null,
  iframeWindow: null,
  connectionId: null
}

export let state = Object.assign({}, initialState)

export function updateState(newState) {
  state = Object.assign({}, state, newState)
}

export function filteredState() {
  const {
    mobileDevice,
    validBrowser,
    currentProvider,
    web3Wallet,
    accessToAccounts,
    walletLoggedIn,
    walletEnabled,
    accountAddress,
    accountBalance,
    minimumBalance,
    userCurrentNetworkId,
    correctNetwork
  } = state

  return {
    mobileDevice,
    validBrowser,
    currentProvider,
    web3Wallet,
    accessToAccounts,
    walletLoggedIn,
    walletEnabled,
    accountAddress,
    accountBalance,
    minimumBalance,
    userCurrentNetworkId,
    correctNetwork
  }
}
