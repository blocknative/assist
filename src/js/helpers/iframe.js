import { updateState, state } from './state'
import { positionElement } from '../views/dom'

import darkModeStyles from '../../css/dark-mode.css'

export function createIframe(browserDocument, assistStyles, style = {}) {
  const { darkMode, css } = style

  const initialIframeContent = `
    <html>
      <head>
        <style>
          ${assistStyles}
        </style>
        <style>
          ${darkMode ? darkModeStyles : ''}
        </style>
        <style>
          ${css || ''}
        </style>
      </head>
      <body></body>
    </html>
  `

  const iframe = positionElement(browserDocument.createElement('iframe'))
  browserDocument.body.appendChild(iframe)
  iframe.style.position = 'fixed'
  iframe.style.height = '0'
  iframe.style.width = '0'
  iframe.style.border = 'none'
  iframe.style.pointerEvents = 'none'
  iframe.style.overflow = 'hidden'
  const iWindow = iframe.contentWindow
  const iDocument = iWindow.document
  iDocument.open()
  iDocument.write(initialIframeContent)
  iDocument.close()

  updateState({ iframe, iframeDocument: iDocument, iframeWindow: iWindow })
}

export function showIframe() {
  state.iframe.style['z-index'] = 999
  state.iframe.style.pointerEvents = 'all'
}

export function hideIframe() {
  state.iframe.style['z-index'] = 'initial'
  state.iframe.style.pointerEvents = 'none'
}

export function resizeIframe({ height, width, transitionHeight }) {
  if (transitionHeight) {
    state.iframe.style.transition = 'height 200ms ease-in-out'
  } else {
    state.iframe.style.transition = 'initial'
  }

  state.iframe.style.height = `${height}px`
  state.iframe.style.width = `${width}px`
}
