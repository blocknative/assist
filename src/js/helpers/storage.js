import { state, updateState } from './state'

export function getItem(item) {
  let storageItem
  try {
    storageItem = window.localStorage && window.localStorage.getItem(item)
  } catch (errorObj) {
    return 'null'
  }
  return storageItem
}

export function storeItem(item, value) {
  try {
    window.localStorage && window.localStorage.setItem(item, value)
  } catch (errorObj) {
    return 'null'
  }
  return 'success'
}

export function removeItem(item) {
  try {
    window.localStorage && window.localStorage.removeItem(item)
  } catch (errorObj) {
    return 'null'
  }
  return 'success'
}

export function storeTransactionQueue() {
  const { transactionQueue } = state
  if (transactionQueue.length > 0) {
    const pendingTransactions = transactionQueue.filter(
      txObj =>
        txObj.transaction.status === 'approved' ||
        txObj.transaction.status === 'pending'
    )
    storeItem('transactionQueue', JSON.stringify(pendingTransactions))
  }
}

export function getTransactionQueueFromStorage() {
  const transactionQueue = getItem('transactionQueue')
  if (transactionQueue) {
    const parsedQueue = JSON.parse(transactionQueue)
    const filteredQueue = parsedQueue.filter(
      txObj => Date.now() - txObj.transaction.startTime < 150000
    )
    updateState({
      transactionQueue: [...filteredQueue, ...state.transactionQueue]
    })
    removeItem('transactionQueue')
  }
}
