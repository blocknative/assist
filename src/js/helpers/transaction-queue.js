import { state, updateState } from './state'

export function addTransactionToQueue(txObject) {
  const { transactionQueue } = state
  return [...transactionQueue, txObject]
}

export function removeTransactionFromQueue(id) {
  const { transactionQueue } = state
  const newQueueState = transactionQueue.filter(
    txObj => txObj.transaction.id !== id
  )

  updateState({ transactionQueue: newQueueState })
}

export function updateTransactionInQueue(id, update) {
  const txObject = getTxObjFromQueue(id)
  txObject.transaction = Object.assign(txObject.transaction, update)

  return txObject
}

export function getTxObjFromQueue(id) {
  const { transactionQueue } = state
  return transactionQueue.find(txObj => txObj.transaction.id === id)
}

export function isDuplicateTransaction({ value, to }) {
  const { transactionQueue } = state

  return Boolean(
    transactionQueue.find(
      txObj => txObj.transaction.value === value && txObj.transaction.to === to
    )
  )
}

export function getTransactionsAwaitingApproval() {
  const { transactionQueue } = state
  return transactionQueue.filter(
    txObj => txObj.transaction.status === 'awaitingApproval'
  )
}

export function isTransactionAwaitingApproval(id) {
  const txObj = getTxObjFromQueue(id)
  return txObj && txObj.transaction.status === 'awaitingApproval'
}
