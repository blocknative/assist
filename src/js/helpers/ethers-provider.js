import { state } from './state'
import { networkName } from './utilities'

let provider

export const getEthersProvider = () => {
  if (!provider) {
    const { ethers } = state.config
    if (typeof window.ethereum !== 'undefined') {
      provider = new ethers.providers.Web3Provider(window.ethereum)
    } else if (typeof window.web3 !== 'undefined') {
      provider = new ethers.providers.Web3Provider(window.web3.currentProvider)
    } else {
      provider = ethers.getDefaultProvider(networkName(state.config.networkId))
    }
  }

  return provider
}

let UncheckedSigner

export const getUncheckedSigner = () => {
  if (!UncheckedSigner) {
    const { ethers } = state.config
    const provider = getEthersProvider()
    class UncheckedJsonRpcSigner extends ethers.Signer {
      constructor(signer) {
        super()
        ethers.utils.defineReadOnly(this, 'signer', signer)
        ethers.utils.defineReadOnly(this, 'provider', signer.provider)
      }

      getAddress() {
        return this.signer.getAddress()
      }

      sendTransaction(transaction) {
        return this.signer.sendUncheckedTransaction(transaction).then(hash => ({
          hash,
          nonce: null,
          gasLimit: null,
          gasPrice: null,
          data: null,
          value: null,
          chainId: null,
          confirmations: 0,
          from: null,
          wait: confirmations =>
            this.provider.waitForTransaction(hash, confirmations)
        }))
      }
    }

    UncheckedSigner = new UncheckedJsonRpcSigner(provider.getSigner())
  }

  return UncheckedSigner
}
