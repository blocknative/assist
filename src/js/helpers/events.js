import { state } from './state'
import eventToUI from '../views/event-to-ui'
import { networkName, timeouts } from './utilities'
import { getTxObjFromQueue } from './transaction-queue'
import { openWebsocketConnection } from './websockets'
import { getItem } from './storage'

export function handleEvent(eventObj, clickHandlers) {
  const { eventCode, categoryCode } = eventObj
  const serverEvent =
    eventCode === 'txPending' ||
    eventCode === 'txConfirmed' ||
    eventCode === 'txFailed'

  // If not a server event then log it
  !serverEvent && lib.logEvent(eventObj)

  // If tx status is 'completed', UI has been already handled
  if (eventCode === 'txConfirmed' || eventCode === 'txConfirmedClient') {
    const txObj = getTxObjFromQueue(eventObj.transaction.id)

    if (txObj.transaction.status === 'completed') {
      return
    }
  }

  // Show UI
  if (state.config && !state.config.headlessMode) {
    eventToUI[categoryCode] &&
      eventToUI[categoryCode][eventCode] &&
      eventToUI[categoryCode][eventCode](eventObj, clickHandlers)
  }
}

// Create event log to be sent to server
export function createEventLog(eventObj) {
  const { dappId, networkId, headlessMode } = state.config
  const { userAgent, version } = state
  const newUser = getItem('_assist_newUser') === 'true'
  return JSON.stringify(
    Object.assign(
      {},
      {
        timeStamp: new Date(),
        dappId,
        version,
        userAgent,
        newUser,
        headlessMode,
        blockchain: {
          system: 'ethereum',
          network: networkName(networkId)
        }
      },
      eventObj
    )
  )
}

// Log events to server
export function logEvent(eventObj) {
  const eventLog = createEventLog(eventObj)
  const { socket, socketConnection } = state

  socketConnection && socket.send(eventLog)

  // Need to check if connection dropped
  // as we don't know until after we try to send a message
  setTimeout(() => {
    if (!state.socketConnection) {
      if (!state.pendingSocketConnection) {
        openWebsocketConnection()
      }

      setTimeout(() => {
        logEvent(eventObj)
      }, timeouts.checkSocketConnection)
    }
  }, timeouts.checkSocketConnection)
}

export const lib = {
  handleEvent,
  createEventLog,
  logEvent
}
