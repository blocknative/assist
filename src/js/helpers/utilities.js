import uuid from 'uuid/v4'
import { handleEvent } from './events'

// Nice time format
export function formatTime(number) {
  const time = new Date(number)
  return time.toLocaleString('en-US', {
    hour: 'numeric',
    minute: 'numeric',
    hour12: true
  })
}

export function timeString(time) {
  const seconds = Math.floor(time / 1000)
  return seconds >= 60 ? `${Math.floor(seconds / 60)} min` : `${seconds} sec`
}

export function capitalize(str) {
  const first = str.slice(0, 1)
  const rest = str.slice(1)
  return `${first.toUpperCase()}${rest}`
}

export function formatNumber(num) {
  const numString = String(num)
  if (numString.includes('+')) {
    const exponent = numString.split('+')[1]
    const precision = Number(exponent) + 1
    return num.toPrecision(precision)
  }
  return num
}

function last(arr) {
  return [...arr].reverse()[0]
}

function takeLast(arr) {
  // mutates original array
  return arr.splice(arr.length - 1, 1)[0]
}

export function first(arr) {
  return arr[0]
}

function takeFirst(arr) {
  // mutates original arr
  return arr.splice(0, 1)[0]
}

export function createTransactionId() {
  return uuid()
}

export function separateArgs(allArgs, argsLength) {
  const allArgsCopy = [...allArgs]
  const args = argsLength ? allArgsCopy.splice(0, argsLength) : []

  const inlineCustomMsgs =
    typeof last(allArgsCopy) === 'object' &&
    last(allArgsCopy).messages &&
    takeLast(allArgsCopy).messages

  const callback =
    typeof last(allArgsCopy) === 'function' && takeLast(allArgsCopy)

  const txObject =
    typeof first(allArgsCopy) === 'object' && first(allArgsCopy) !== null
      ? takeFirst(allArgsCopy)
      : {}

  const defaultBlock = first(allArgsCopy)

  return {
    callback,
    args,
    txObject,
    defaultBlock,
    inlineCustomMsgs
  }
}

export function getOverloadedMethodKeys(inputs) {
  return inputs.map(input => input.type).join(',')
}

export function assistLog(log) {
  console.log('Assist:', log) // eslint-disable-line no-console
}

export function extractMessageFromError(message) {
  const str = message.split('"message":')[1]
  return str.split('"')[1]
}

export function eventCodeToType(eventCode) {
  switch (eventCode) {
    case 'txRequest':
    case 'txPending':
    case 'txSent':
      return 'progress'
    case 'txSendFail':
    case 'txStall':
    case 'txFailed':
    case 'nsfFail':
    case 'txRepeat':
    case 'txAwaitingApproval':
    case 'txConfirmReminder':
      return 'failed'
    case 'txConfirmed':
    case 'txConfirmedClient':
      return 'complete'
    default:
      return undefined
  }
}

export function eventCodeToStep(eventCode) {
  switch (eventCode) {
    case 'mobileBlocked':
      return 'mobile'
    case 'browserFail':
      return 'browser'
    case 'welcomeUser':
      return 0
    case 'walletFail':
      return 1
    case 'walletLogin':
    case 'walletLoginEnable':
      return 2
    case 'walletEnable':
      return 3
    case 'networkFail':
    case 'contractQueryFail':
      return 4
    case 'nsfFail':
      return 5
    case 'newOnboardComplete':
      return 6
    default:
      return undefined
  }
}

export function networkName(id) {
  switch (Number(id)) {
    case 1:
      return 'main'
    case 3:
      return 'ropsten'
    case 4:
      return 'rinkeby'
    case 42:
      return 'kovan'
    case 'localhost':
      return 'localhost'
    default:
      return 'local'
  }
}

export const timeouts = {
  checkSocketConnection: 250,
  waitForResponse: 100,
  txConfirmReminder: 20000,
  txStall: 30000,
  changeUI: 305,
  localhostNetworkCheck: 300,
  removeElement: 300,
  endOfEventQueue: 0,
  hideElement: 200,
  showElement: 120,
  autoRemoveNotification: 4000,
  pollForReceipt: 1000
}

export function stepToImageKey(step) {
  switch (step) {
    case 0:
      return 'welcome'
    case 6:
      return 'complete'
    default:
      return null
  }
}

export function handleError(handlers = {}) {
  return errorObj => {
    const { callback, reject, resolve } = handlers

    if (callback) {
      callback(errorObj)
      resolve()

      return
    }

    reject(errorObj)
  }
}

export function handleWeb3Error(errorObj) {
  const { message } = errorObj
  handleEvent({
    eventCode: 'errorLog',
    categoryCode: 'web3',
    reason: message || errorObj
  })
}
