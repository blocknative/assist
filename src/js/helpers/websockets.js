import { state, updateState } from './state'
import { handleEvent } from './events'
import { storeItem, getItem } from './storage'
import { assistLog } from './utilities'
import {
  updateTransactionInQueue,
  getTxObjFromQueue,
  removeTransactionFromQueue
} from './transaction-queue'

// Create websocket connection
export function openWebsocketConnection() {
  updateState({ pendingSocketConnection: true })

  let socket
  try {
    socket = new WebSocket('wss://api.blocknative.com/v0')
  } catch (errorObj) {
    assistLog(errorObj)
  }

  socket.addEventListener('message', handleSocketMessage)
  socket.addEventListener('close', () =>
    updateState({ socketConnection: false })
  )
  socket.addEventListener('error', () => {
    updateState({ pendingSocketConnection: false })
  })
  socket.addEventListener('open', () => {
    updateState({
      socket,
      socketConnection: true,
      pendingSocketConnection: false
    })

    handleEvent({
      categoryCode: 'initialize',
      eventCode: 'checkDappId',
      connectionId: getItem('connectionId')
    })
  })
}

// Handle in coming socket messages
export function handleSocketMessage(msg) {
  const { status, reason, event, connectionId } = JSON.parse(msg.data)
  const { validApiKey, supportedNetwork } = state
  if (!validApiKey || !supportedNetwork) {
    return
  }

  // handle any errors from the server
  if (status === 'error') {
    if (
      reason.includes('not a valid API key') &&
      event.eventCode !== 'initFail'
    ) {
      updateState({ validApiKey: false })

      handleEvent({
        eventCode: 'initFail',
        categoryCode: 'initialize',
        reason
      })

      const errorObj = new Error(reason)
      errorObj.eventCode = 'initFail'
      throw errorObj
    }

    if (
      reason.includes('network not supported') &&
      event.eventCode !== 'initFail'
    ) {
      updateState({ supportedNetwork: false })

      handleEvent({
        eventCode: 'initFail',
        categoryCode: 'initialize',
        reason
      })

      const errorObj = new Error(reason)
      errorObj.eventCode = 'initFail'
      throw errorObj
    }
  }

  if (status === 'ok' && event && event.eventCode === 'checkDappId') {
    handleEvent({ eventCode: 'initSuccess', categoryCode: 'initialize' })
    updateState({ validApiKey: true, supportedNetwork: true })
  }

  if (
    connectionId &&
    (!state.connectionId || connectionId !== state.connectionId)
  ) {
    storeItem('connectionId', connectionId)
    updateState({ connectionId })
  }

  if (event && event.transaction) {
    const { transaction, eventCode } = event
    let txObj

    switch (transaction.status) {
      case 'pending':
        txObj = updateTransactionInQueue(transaction.id, {
          status: 'pending',
          nonce: transaction.nonce
        })

        handleEvent({
          eventCode: eventCode === 'txPool' ? 'txPending' : eventCode,
          categoryCode: 'activeTransaction',
          transaction: txObj.transaction,
          contract: txObj.contract,
          inlineCustomMsgs: txObj.inlineCustomMsgs
        })
        break
      case 'confirmed':
        txObj = getTxObjFromQueue(transaction.id)

        if (txObj.transaction.status === 'confirmed') {
          txObj = updateTransactionInQueue(transaction.id, {
            status: 'completed'
          })
        } else {
          txObj = updateTransactionInQueue(transaction.id, {
            status: 'confirmed'
          })
        }

        handleEvent({
          eventCode: 'txConfirmed',
          categoryCode: 'activeTransaction',
          transaction: txObj.transaction,
          contract: txObj.contract,
          inlineCustomMsgs: txObj.inlineCustomMsgs
        })

        if (txObj.transaction.status === 'completed') {
          removeTransactionFromQueue(transaction.id)
        }

        break
      case 'failed':
        txObj = updateTransactionInQueue(transaction.id, { status: 'failed' })

        handleEvent({
          eventCode: 'txFailed',
          categoryCode: 'activeTransaction',
          transaction: txObj.transaction,
          contract: txObj.contract,
          inlineCustomMsgs: txObj.inlineCustomMsgs
        })

        removeTransactionFromQueue(transaction.id)
        break
      default:
    }
  }
}
