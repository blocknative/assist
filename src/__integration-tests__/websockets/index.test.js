import { WebSocket, Server } from 'mock-socket'

import da from '~/js'
import { state, updateState, initialState } from '~/js/helpers/state'
import * as websockets from '~/js/helpers/websockets'
import * as utilities from '~/js/helpers/utilities'
import * as events from '~/js/helpers/events'

const socketUrl = 'wss://api.blocknative.com/v0'

describe('a websocket connection is requested', () => {
  let mockServer
  const config = { dappId: '123', networkId: '1' }
  beforeEach(() => {
    da.init(config)
    mockServer = new Server(socketUrl)
  })
  afterEach(() => {
    mockServer.close()
  })
  test('state.pendingSocketConnection is set to true', () => {
    websockets.openWebsocketConnection()
    expect(state.pendingSocketConnection).toEqual(true)
  })
  describe('socket creation succeeds', () => {
    let handleEventSpy
    const connectionId = '123'
    beforeEach(() => {
      handleEventSpy = jest
        .spyOn(events, 'handleEvent')
        .mockImplementation(() => {})
      window.localStorage.setItem('connectionId', connectionId)
      websockets.openWebsocketConnection()
      mockServer.emit('open')
    })
    afterEach(() => {
      handleEventSpy.mockRestore()
    })
    test('state.pendingSocketConnection set to false', () => {
      expect(state.pendingSocketConnection).toEqual(false)
    })
    test('state.socketConnection set to true', () => {
      expect(state.socketConnection).toEqual(true)
    })
    test('state.socket refs the socket', () => {
      expect(state.socket.url).toEqual(socketUrl)
    })
    test('event checkDappId is emitted with correct payload', () => {
      expect(handleEventSpy).toBeCalledWith({
        categoryCode: 'initialize',
        eventCode: 'checkDappId',
        connectionId
      })
    })
  })
  describe('socket refuses to connect', () => {
    beforeEach(() => {
      websockets.openWebsocketConnection()
      mockServer.emit('error')
    })
    test('state.pendingSocketConnection is set to false', () => {
      expect(state.pendingSocketConnection).toEqual(false)
    })
  })
  describe('socket creation throws an error', () => {
    beforeEach(() => {
      global.WebSocket = null
    })
    afterEach(() => {
      global.WebSocket = WebSocket
    })
    test('assistLog should be called with the error', () => {
      const assistLogSpy = jest
        .spyOn(utilities, 'assistLog')
        .mockImplementation(() => {})
      expect(() => {
        websockets.openWebsocketConnection()
      }).toThrow()
      expect(assistLogSpy).toHaveBeenCalled()
      assistLogSpy.mockRestore()
    })
  })
})

describe('assist is connected to a websocket', () => {
  const config = { dappId: '123', networkId: '1' }
  let mockServer
  beforeEach(() => {
    da.init(config)
    mockServer = new Server(socketUrl)
    websockets.openWebsocketConnection()
    mockServer.emit('open')
  })
  afterEach(() => {
    mockServer.close()
  })
  describe('receives a message with a connectionId', () => {
    const payload = { connectionId: '123' }
    test(`if state.connectionId is null it should be set in state and localStorage`, () => {
      mockServer.emit('message', JSON.stringify(payload))
      expect(state.connectionId).toEqual(payload.connectionId)
      expect(window.localStorage.getItem('connectionId')).toEqual(
        payload.connectionId
      )
    })
    test(`if state.connectionId !== connectionId from the server, update with val from server`, () => {
      updateState({ connectionId: '456' })
      expect(state.connectionId).toEqual('456')
      mockServer.emit('message', JSON.stringify(payload))
      expect(state.connectionId).toEqual(payload.connectionId)
      expect(window.localStorage.getItem('connectionId')).toEqual(
        payload.connectionId
      )
    })
  })
  describe('received message: [not a valid API key]', () => {
    const payload = {
      status: 'error',
      reason: 'not a valid API key',
      event: { eventCode: 'mockEvent' }
    }
    test('the expected error should be thrown', () => {
      expect(() => {
        mockServer.emit('message', JSON.stringify(payload))
      }).toThrowErrorMatchingSnapshot()
    })
    test('state.validApiKey should be set to false', () => {
      expect(() => {
        mockServer.emit('message', JSON.stringify(payload))
      }).toThrow()
      expect(state.validApiKey).toEqual(false)
    })
    test('correct initFail event should be emitted', () => {
      const handleEventSpy = jest
        .spyOn(events, 'handleEvent')
        .mockImplementation(() => {})
      expect(() => {
        mockServer.emit('message', JSON.stringify(payload))
      }).toThrow()
      expect(handleEventSpy).toHaveBeenCalledWith({
        eventCode: 'initFail',
        categoryCode: 'initialize',
        reason: payload.reason
      })
      handleEventSpy.mockRestore()
    })
  })
  describe('received message: [network not supported]', () => {
    const payload = {
      status: 'error',
      reason: 'network not supported',
      event: { eventCode: 'mockEvent' }
    }
    test('the expected error should be thrown', () => {
      expect(() => {
        mockServer.emit('message', JSON.stringify(payload))
      }).toThrowErrorMatchingSnapshot()
    })
    test('state.validApiKey should be set to false', () => {
      expect(() => {
        mockServer.emit('message', JSON.stringify(payload))
      }).toThrow()
      expect(state.supportedNetwork).toEqual(false)
    })
    test('correct initFail event should be emitted', () => {
      const handleEventSpy = jest
        .spyOn(events, 'handleEvent')
        .mockImplementation(() => {})
      expect(() => {
        mockServer.emit('message', JSON.stringify(payload))
      }).toThrow()
      expect(handleEventSpy).toHaveBeenCalledWith({
        eventCode: 'initFail',
        categoryCode: 'initialize',
        reason: payload.reason
      })
      handleEventSpy.mockRestore()
    })
  })
  describe('received message: [checkDappId]', () => {
    const payload = { status: 'ok', event: { eventCode: 'checkDappId' } }
    beforeEach(() => {
      mockServer.emit('message', JSON.stringify(payload))
    })
    test('state.validApiKey should be set to true', () => {
      expect(state.validApiKey).toEqual(true)
    })
    test('state.supportedNetwork should be set to true', () => {
      expect(state.supportedNetwork).toEqual(true)
    })
  })
  describe('received message with a pending transaction', () => {
    let payload
    let handleEventSpy
    const transaction = {
      status: 'pending',
      nonce: 123,
      id: 'some-id'
    }
    const txObj = {
      transaction,
      eventCode: 'some-event',
      contract: 'some-contract',
      inlineCustomMsgs: { '1': 'msg' }
    }
    beforeEach(() => {
      const existingTx = {
        ...transaction,
        status: 'submitted'
      }
      updateState({ transactionQueue: [{ ...txObj, transaction: existingTx }] })
      handleEventSpy = jest
        .spyOn(events, 'handleEvent')
        .mockImplementation(() => {})
      mockServer.emit('message', JSON.stringify(payload))
    })
    afterEach(() => {
      handleEventSpy.mockRestore()
    })
    payload = { event: { transaction } }
    test('tx status in transactionQueue should be set to pending', () => {
      expect(state.transactionQueue[0].transaction.status).toEqual('pending')
    })
    test('the expected event should be emitted', () => {
      expect(handleEventSpy).toHaveBeenCalledWith({
        eventCode: payload.eventCode,
        categoryCode: 'activeTransaction',
        transaction: txObj.transaction,
        contract: txObj.contract,
        inlineCustomMsgs: txObj.inlineCustomMsgs
      })
    })
    describe('eventCode is txPool', () => {
      beforeAll(() => {
        payload = {
          event: { transaction, eventCode: 'txPool' }
        }
      })
      test(`emitted event should have eventCode 'txPending'`, () => {
        expect(handleEventSpy).toHaveBeenCalledWith({
          eventCode: 'txPending',
          categoryCode: 'activeTransaction',
          transaction: txObj.transaction,
          contract: txObj.contract,
          inlineCustomMsgs: txObj.inlineCustomMsgs
        })
      })
    })
  })
  describe('received message with a confirmed transaction', () => {
    const transaction = {
      status: 'confirmed',
      nonce: 123,
      id: 'some-id'
    }
    const txObj = {
      transaction,
      eventCode: 'some-event',
      contract: 'some-contract',
      inlineCustomMsgs: { '1': 'msg' }
    }
    let existingTxStatus = 'pending'
    let handleEventSpy
    beforeEach(() => {
      handleEventSpy = jest
        .spyOn(events, 'handleEvent')
        .mockImplementation(() => {})
      const existingTx = {
        ...transaction,
        status: existingTxStatus
      }
      updateState({ transactionQueue: [{ ...txObj, transaction: existingTx }] })
    })
    afterEach(() => {
      handleEventSpy.mockRestore()
    })
    const payload = { event: { transaction } }
    describe('tx status is confirmed', () => {
      beforeAll(() => {
        existingTxStatus = 'confirmed'
      })
      beforeEach(() => {
        mockServer.emit('message', JSON.stringify(payload))
      })
      test(`tx should be removed from the txQueue`, () => {
        expect(
          state.transactionQueue.find(
            txObj => txObj.transaction.id === 'some-id'
          )
        ).toBeFalsy()
      })
      test(`correct event should be emitted with tx status 'completed'`, () => {
        expect(handleEventSpy).toHaveBeenCalledWith({
          eventCode: 'txConfirmed',
          categoryCode: 'activeTransaction',
          transaction: { ...txObj.transaction, status: 'completed' },
          contract: txObj.contract,
          inlineCustomMsgs: txObj.inlineCustomMsgs
        })
      })
    })
    describe('tx status is NOT confirmed', () => {
      beforeAll(() => {
        existingTxStatus = 'some-status'
      })
      beforeEach(() => {
        mockServer.emit('message', JSON.stringify(payload))
      })
      test(`tx should remain in the txQueue`, () => {
        expect(
          state.transactionQueue.find(
            txObj => txObj.transaction.id === 'some-id'
          )
        ).toBeTruthy()
      })
      test(`correct event should be emitted with tx status 'confirmed'`, () => {
        expect(handleEventSpy).toHaveBeenCalledWith({
          eventCode: 'txConfirmed',
          categoryCode: 'activeTransaction',
          transaction: { ...txObj.transaction, status: 'confirmed' },
          contract: txObj.contract,
          inlineCustomMsgs: txObj.inlineCustomMsgs
        })
      })
    })
  })
  describe('received message with a failed transaction', () => {
    let handleEventSpy
    const transaction = {
      status: 'failed',
      nonce: 123,
      id: 'some-id'
    }
    const txObj = {
      transaction,
      eventCode: 'some-event',
      contract: 'some-contract',
      inlineCustomMsgs: { '1': 'msg' }
    }
    const existingTx = {
      ...transaction,
      status: 'some-status'
    }
    const payload = { event: { transaction } }
    beforeEach(() => {
      updateState({ transactionQueue: [{ ...txObj, transaction: existingTx }] })
      handleEventSpy = jest
        .spyOn(events, 'handleEvent')
        .mockImplementation(() => {})
      mockServer.emit('message', JSON.stringify(payload))
    })
    afterEach(() => {
      handleEventSpy.mockRestore()
    })
    test(`tx should be removed from the txQueue`, () => {
      expect(
        state.transactionQueue.find(txObj => txObj.transaction.id === 'some-id')
      ).toBeFalsy()
    })
    test(`correct event should be emitted with tx status 'failed'`, () => {
      expect(handleEventSpy).toHaveBeenCalledWith({
        eventCode: 'txFailed',
        categoryCode: 'activeTransaction',
        transaction: { ...txObj.transaction, status: 'failed' },
        contract: txObj.contract,
        inlineCustomMsgs: txObj.inlineCustomMsgs
      })
    })
  })
})

afterEach(() => {
  document.body.innerHTML = ''
  updateState(initialState)
  window.localStorage.clear()
  jest.clearAllMocks()
})
