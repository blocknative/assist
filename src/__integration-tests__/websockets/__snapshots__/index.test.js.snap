// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assist is connected to a websocket received message: [network not supported] the expected error should be thrown 1`] = `"network not supported"`;

exports[`assist is connected to a websocket received message: [not a valid API key] the expected error should be thrown 1`] = `"not a valid API key"`;
